### DISCUSSION TITLE
Lambda Deploy

### DISCUSSION THREAD

**Discussion by rbgfloripa** [remix-run/remix#4678] [Q&A]
Hi all,

Is there any way to deploy the application in an aws lambda WITHOUT Architect?
Because my deployment process is completely separate and automated. I don't have access to the AWS account.
I think I would have to generate a ZIP file of the application to run in a lambda or something like this.

Thanks.

**Comment by rbgfloripa** [comment]
I've tried it in several ways with no success.
I'm stuck with this problem and can't move forward.
I will probably need to abandon the idea of ​​using remix in this project.

**Reply by vinarius** [reply]
same, nearly 2 years later :(

**Comment by MattyBalaam** [comment]
I’m in a similar situation. I want to bring in Remix to our stack, which in terms of frontend apps is currently PWA hosted in S3 and our company is going to be staying AWS for quite a long time. I personally need a bit of hand holding around how to use edge computing in AWS, and would rather not need to briing in yet another dependency (Architect) if I can help it. 

I see there is something called lambda@edge - would this be the correct thing to look into? https://aws.amazon.com/lambda/edge/  Or are the simple Cloudfront Functions the way to go? https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/edge-functions.html

**Comment by rbgfloripa** [comment]
I'm trying to run the application on a simple lambda for now.

**Reply by thomasfr** [reply]
Have you been successful in running it on Lambda? If yes, what did you need to do?

**Comment by devraj** [comment]
I have been wanting to achieve the same thing without adding another dependency like Architect. We use a Kubernetes (deployed using Terraform) based setup which provides us some options:

- [Deploying Remix apps to Kubernetes?](https://shipa.io/deploying-remix-apps-to-kubernetes/)
- Found this [discussion](https://github.com/remix-run/remix/discussions/5598) which provides a links to [Deploying Remix to Kubernetes](https://kaibun.net/articles/2023-02-25-deploying-remix-to-kubernetes) written by @chikamichi
- @BenMcH maintains the [techno-stack](https://github.com/BenMcH/techno-stack) while I wont' want to buy into the stack it will be well worth looking at how it's been put together.

We work with Python on the server and have PWAs at the moment, which I am looking to replace with remix based apps

**Reply by BenMcH** [reply]
Hi, thanks for the shout out but I haven't been working on Web apps recently, so the techno stack is woefully out of date. The kubernetes deploy files are there though and should still work 🙂

**Reply by devraj** [reply]
Also tagging this issue reported by @nicolaserny 
https://github.com/remix-run/remix/issues/4176

**Comment by fones** [comment]
Hi guys, did someone tried to use custom lambda container image for this? @rbgfloripa @devraj 

You can deploy lambda as docker image with some limitations.  https://docs.aws.amazon.com/lambda/latest/dg/images-create.html

**Reply by thomasfr** [reply]
We are using the Lambda Web Adapter as a Lambda extension and using Docker as a deployment container.

**Reply by thomasfr** [reply]
It’s this project: https://github.com/awslabs/aws-lambda-web-adapter

**Reply by fones** [reply]
@thomasfr and you deploy remix with success?

**Reply by thomasfr** [reply]
Yes! We use it at AWS internally

**Reply by thomasfr** [reply]
But you don’t have to use docker to use the Lambda web adapter. You can also use it without docker. Alternatively, you can also do the Request Handling in the entry.server.ts file. But the LWA makes that unnecessary.

**Reply by jarimustonen** [reply]
@thomasfr Would it be possible to have a bit more in-depth instructions? I noticed that there are lot of examples but Remix and Remix in ZIP are missing. Also, it would be great if the example would use Vite as Remix is in the process of migrating to Vite.

Edit: I went and added an issue for this: https://github.com/awslabs/aws-lambda-web-adapter/issues/387

As you are using it internally, it would really make sense to also embrace this externally.

**Comment by mogadanez** [comment]
here is exampe:
https://github.com/shamsup/remix-starter-serverless
it still uses 
import { createRequestHandler } from "@remix-run/architect";

but not full architect package

you can inspire how createRequestHandler looks like and migrate it to your codebase

**Reply by gotgenes** [reply]
This project uses a version of Remix that predates the current Vite architecture, which [became the default in v2.7.0](https://remix.run/blog/remix-vite-stable).

I found relevant discussion at #8836.

**Comment by rreeves8** [comment]
Opening this discussion back up.

I'm deploying a remix app on lambda, but I'm going through express first 

```
import serverlessExpress from "@codegenie/serverless-express";
```
..more code
```
if (!isProd) {
  app.listen(80, "127.0.0.2", () => {
    console.log("App listening on http://127.0.0.2:80");
  });
}
console.log("finished loading server");

export const handler = serverlessExpress({
  app,
});
```

```
FROM public.ecr.aws/lambda/nodejs:20


COPY package*.json ${LAMBDA_TASK_ROOT}
COPY node_modules ${LAMBDA_TASK_ROOT}/node_modules
COPY build ${LAMBDA_TASK_ROOT}/build

#for mongo tls, its on a PE
COPY cert.pem ${LAMBDA_TASK_ROOT}

COPY app.js ${LAMBDA_TASK_ROOT}

ENV NODE_ENV=production

CMD [ "app.handler" ]
```

None of my actions work and my loaders are broken for some reason. does anyone know if this setup will actually work ? I think the conversion for the request isn't functioning ? I'm getting this error for some routes SyntaxError: The string did not match the expected pattern.

**Comment by vsnig** [comment]
https://serverlessup.com/ramblings/remix-&-cdk-fun-08-11-2022

This post can be helpful (not mine). It describes deployment to CloudFront->APIGW->Lambda with CDK.

Personally I don't understand what Remix has to do with some Architect and why it is even being discussed. I never heard of it before nor after except in context of Remix. Is it some "partner project" or what, I don't understand.

The lack of attention from Remix team to AWS deployment upsets me.

**Reply by MattyBalaam** [reply]
Very useful article.

**Comment by gotgenes** [comment]
I published [a repository to demonstrate deploying to AWS Lambda using CDK](https://github.com/gotgenes/remix-aws-lambda-app), based on [the solution I described in a different Discussion](https://github.com/remix-run/remix/discussions/8836#discussioncomment-9795687).

It has some similarities to @vsnig's [solution above](https://github.com/remix-run/remix/discussions/4678#discussioncomment-10878893) but has a simpler architecture.

**Reply by vsnig** [reply]
Just for the record, the solution is not mine, i just posted a link, thanks. As for me, in my projects I also use Lambda URL instead of APIGW.
I also noticed you allow traffic both via CloudFront distribution url and Lambda URL, which can be undesirable. If you want to allow only Cloudfront and block requests coming directly to lambda url, the solution is in that blog post that you mentioned (`x-origin-verify` header). 
Also, `origins.S3Origin`, is deprecated since not so long ago, I use `origins.S3BucketOrigin.withOriginAccessControl(staticBucket)`.
Good luck!

**Reply by gotgenes** [reply]
Very helpful; thanks, @vsnig!

**Reply by gotgenes** [reply]
> Also, `origins.S3Origin`, is deprecated since not so long ago, I use `origins.S3BucketOrigin.withOriginAccessControl(staticBucket)`.

I tried this but ran into aws/aws-cdk#31462. I will have to investigate further. It seems like I will have to put the assets bucket into the same stack as the CloudFront distribution. That's probably acceptable in my case, but surprising.

> I also noticed you allow traffic both via CloudFront distribution url and Lambda URL, which can be undesirable. If you want to allow only Cloudfront and block requests coming directly to lambda url, the solution is in that blog post that you mentioned (x-origin-verify header).

I looked alternatively at [Origin Access Control for Lambda](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-lambda.html), but this blocks `POST` requests from going through straight from the browser. [Some workarounds exist](https://repost.aws/questions/QUAE2ba-zHTHqwVffGkkzL3Q/how-to-safely-call-lambda-function-urls-from-cloudfront-including-post#ANluRLOEm9SumrsbiQ6ZD7lg), but perhaps these indicate you either need to be comfortable with having a publicly accessible Lambda URL, or switch to putting API Gateway in between CloudFront and the Remix Lambda. I'll have to think more about my approach.

Thanks again for pointing these out, @vsnig.

**Reply by curran** [reply]
Regarding the `S3Origin` being deprecated, I ran into this issue as well and had to do quite a lot of extra work to use the alternative API. Here's what I have that is fully working, it may be useful:

```js

import { Construct } from "constructs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as apigw from "aws-cdk-lib/aws-apigateway";
import { HostedZone, ARecord, RecordTarget } from "aws-cdk-lib/aws-route53";
import {
  Certificate,
  CertificateValidation,
} from "aws-cdk-lib/aws-certificatemanager";
import * as targets from "aws-cdk-lib/aws-route53-targets";
import * as iam from "aws-cdk-lib/aws-iam";
import * as secretsmanager from "aws-cdk-lib/aws-secretsmanager";
import {
  Distribution,
  ViewerProtocolPolicy,
  CachePolicy,
  AllowedMethods,
} from "aws-cdk-lib/aws-cloudfront";
import { HttpOrigin, S3BucketOrigin } from "aws-cdk-lib/aws-cloudfront-origins";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as s3deploy from "aws-cdk-lib/aws-s3-deployment";
import { Duration, RemovalPolicy, Stack } from "aws-cdk-lib";

interface RemixAppConstructProps {
  domainName: string;
  subDomain: string;
}

export class RemixAppConstruct extends Construct {
  constructor(scope: Construct, id: string, props: RemixAppConstructProps) {
    super(scope, id);

    const { domainName, subDomain } = props;
    const fullDomainName = `${subDomain}.${domainName}`;

    // Use the provided hosted zone or look it up
    const hostedZone = HostedZone.fromLookup(this, "HostedZone", {
      domainName: domainName,
    });

    const certificate = new Certificate(this, "ApiCertificate", {
      domainName: fullDomainName,
      validation: CertificateValidation.fromDns(hostedZone),
    });

    // Define the IAM Role with an explicit physical name
    const remixAppRole = new iam.Role(this, "RemixAppRole", {
      roleName: "RemixAppRole", // Explicit physical name
      assumedBy: new iam.ServicePrincipal("lambda.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSLambdaBasicExecutionRole"
        ),
      ],
    });

    // Grant Lambda outbound internet access permissions
    remixAppRole.addToPolicy(
      new iam.PolicyStatement({
        actions: ["ec2:CreateNetworkInterface", "ec2:DeleteNetworkInterface"],
        resources: ["*"],
      })
    );

    // Grant permissions to read the secret
    const secret = secretsmanager.Secret.fromSecretNameV2(
      this,
      "SuperDuperSecret",
      "SECRET_ENV_VARS"
    );
    secret.grantRead(remixAppRole);

    // Define the Lambda function
    const remixAppFunction = new lambda.Function(this, "RemixAppFunction", {
      runtime: lambda.Runtime.NODEJS_20_X,
      handler: "index.handler",
      code: lambda.Code.fromAsset("packages/remix-app"),
      timeout: Duration.minutes(15), // Set timeout to 15 minutes
      environment: {},
      role: remixAppRole,
    });

    const api = new apigw.RestApi(this, "RemixAppAPI", {
      restApiName: "RemixApp Service",
      description: "This service hosts the Remix app.",
      deployOptions: {
        stageName: "prod",
      },
    });

    const apiDomainName = new apigw.DomainName(this, "ApiDomainName", {
      domainName: fullDomainName,
      certificate: certificate,
      endpointType: apigw.EndpointType.REGIONAL,
      securityPolicy: apigw.SecurityPolicy.TLS_1_2,
    });

    new apigw.BasePathMapping(this, "BasePathMapping", {
      domainName: apiDomainName,
      restApi: api,
      basePath: "",
    });

    const integration = new apigw.LambdaIntegration(remixAppFunction, {
      proxy: true,
    });

    const proxyResource = api.root.addResource("{proxy+}");
    proxyResource.addMethod("ANY", integration);

    // Create an S3 bucket for the static assets
    const staticAssetsBucket = new s3.Bucket(this, "StaticAssetsBucket", {
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
      objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_ENFORCED, // Required for OAC
    });

    // Deploy static assets to the S3 bucket
    new s3deploy.BucketDeployment(this, "DeployStaticAssets", {
      sources: [s3deploy.Source.asset("packages/remix-app/build/client")],
      destinationBucket: staticAssetsBucket,
    });

    // Create an S3 origin using Origin Access Control (OAC)
    const s3Origin = S3BucketOrigin.withOriginAccessControl(staticAssetsBucket);

    // Set up CloudFront distribution with both API Gateway and S3 origins
    const distribution = new Distribution(this, "ApiCloudFrontDistribution", {
      defaultBehavior: {
        origin: new HttpOrigin(
          `${api.restApiId}.execute-api.${Stack.of(this).region}.amazonaws.com`,
          {
            originPath: "/prod", // Ensure this matches the API Gateway stage
          }
        ),
        viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        cachePolicy: CachePolicy.CACHING_DISABLED, // For dynamic content
        allowedMethods: AllowedMethods.ALLOW_ALL,
      },
      domainNames: [fullDomainName],
      certificate: certificate,
      additionalBehaviors: {
        "favicon.ico": {
          origin: s3Origin,
          viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
          cachePolicy: CachePolicy.CACHING_OPTIMIZED,
        },
        "logo-*": {
          origin: s3Origin,
          viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
          cachePolicy: CachePolicy.CACHING_OPTIMIZED,
        },
        "/assets/*": {
          origin: s3Origin,
          viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
          cachePolicy: CachePolicy.CACHING_OPTIMIZED,
        },
      },
    });

    // Update the Route 53 ARecord to point to the CloudFront distribution
    new ARecord(this, "ApiAliasRecord", {
      zone: hostedZone,
      recordName: subDomain,
      target: RecordTarget.fromAlias(
        new targets.CloudFrontTarget(distribution)
      ),
    });
  }
}
```

**Reply by vsnig** [reply]
> > Also, `origins.S3Origin`, is deprecated since not so long ago, I use `origins.S3BucketOrigin.withOriginAccessControl(staticBucket)`.
> 
> I tried this but ran into [aws/aws-cdk#31462](https://github.com/aws/aws-cdk/issues/31462). I will have to investigate further. It seems like I will have to put the assets bucket into the same stack as the CloudFront distribution. That's probably acceptable in my case, but surprising.
> 
Yeah, I also ran into this issue with a project where I had to have the bucket in a separate stack, and ended up using `S3BucketOrigin.withOriginAccessIdentity` which is not recommended but works. As for static bucket in Remix deployement, I have it in the same stack as the distribution, similar to the code posted above.

**Reply by vsnig** [reply]
> I looked alternatively at [Origin Access Control for Lambda](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-lambda.html), but this blocks `POST` requests from going through straight from the browser. [Some workarounds exist](https://repost.aws/questions/QUAE2ba-zHTHqwVffGkkzL3Q/how-to-safely-call-lambda-function-urls-from-cloudfront-including-post#ANluRLOEm9SumrsbiQ6ZD7lg), but perhaps these indicate you either need to be comfortable with having a publicly accessible Lambda URL, or switch to putting API Gateway in between CloudFront and the Remix Lambda. I'll have to think more about my approach.

Have you looked at `x-origin-verify` option? It works with any request and doesn't add very much complexity. Basically, you pass some meaningless passphrase from Cloudfront to Lambda via `x-origin-verify` header, then in Lambda, check the presence and the value of the header. If matches - which means that the request properly came through Cloudfront - proceed, if not - throw error.

CDK:
```typescript
// generate passphrase
const xOriginVerify = cdk.Names.uniqueId(this)
...
const serverFunction = new NodejsFunction(this, 'RemixFunction', {
  ...
  // make lambda know the value so it can verify value passed from Cloudfront
  environment: { X_ORIGIN_VERIFY: xOriginVerify }
}
...
new cloudfront.Distribution(this, `remixDistribution`, {
  defaultBehavior: {
    origin: new origins.HttpOrigin(serverFunctionUrlDomain, {
        customHeaders: { 'x-origin-verify': xOriginVerify },
     }),    
    cachePolicy: {
      headerBehavior: cloudfront.CacheHeaderBehavior.allowList(
        'x-origin-verify',
     ),
    ...
}
```

Lambda handler:
```typescript
....
export const handler: RequestHandler = (event, context, callback) => {
  const xOriginVerify = process.env.X_ORIGIN_VERIFY

  if (xOriginVerify && event.headers['x-origin-verify'] !== xOriginVerify) {
    return Promise.resolve({
      statusCode: 401,
    })
  }

  return createRequestHandler({
    build,
    mode: process.env.NODE_ENV,
  })(event, context, callback)
}
```

**Comment by curran** [comment]
Thank you @gotgenes for sharing your solution! I studied it and learned a lot about how CDK projects can be organized.

This alternative approach is working well for me in Lambda:

```js
const serverlessExpress = require("@codegenie/serverless-express");
const express = require("express");
const { createRequestHandler } = require("@remix-run/express");

const app = express();

let build;
exports.handler = async (event, context) => {
  if (!build) {
    build = await import("./build/server/index.js");
  }

  app.all("*", createRequestHandler({ build }));

  return serverlessExpress({ app })(event, context);
};

```

I tried with Architect but that seemed too heavyweight, and actually made `node_modules` too large for Lambda. This gets around that by using `@remix-run/express`, which seems pretty widely used and is even featured in their [quickstart: bring-your-own-server](https://remix.run/docs/en/main/start/quickstart#bring-your-own-server). That's paired with [@codegenie/serverless-express](https://www.npmjs.com/package/@codegenie/serverless-express) which also seems pretty reputable and widely used. This setup does not depend on the ` "remix-aws": "^1.3.0-beta.1",` dependency found in the [previously shared solution](https://github.com/gotgenes/remix-aws-lambda-app/blob/main/site/package.json#L25C4-L25C34) by @gotgenes.

**Reply by pimentellima** [reply]
> Thank you @gotgenes for sharing your solution! I studied it and learned a lot about how CDK projects can be organized.
> 
> This alternative approach is working well for me in Lambda:
> 
> ```js
> const serverlessExpress = require("@codegenie/serverless-express");
> const express = require("express");
> const { createRequestHandler } = require("@remix-run/express");
> 
> const app = express();
> 
> let build;
> exports.handler = async (event, context) => {
>   if (!build) {
>     build = await import("./build/server/index.js");
>   }
> 
>   app.all("*", createRequestHandler({ build }));
> 
>   return serverlessExpress({ app })(event, context);
> };
> ```
> 
> I tried with Architect but that seemed too heavyweight, and actually made `node_modules` too large for Lambda. This gets around that by using `@remix-run/express`, which seems pretty widely used and is even featured in their [quickstart: bring-your-own-server](https://remix.run/docs/en/main/start/quickstart#bring-your-own-server). That's paired with [@codegenie/serverless-express](https://www.npmjs.com/package/@codegenie/serverless-express) which also seems pretty reputable and widely used. This setup does not depend on the ` "remix-aws": "^1.3.0-beta.1",` dependency found in the [previously shared solution](https://github.com/gotgenes/remix-aws-lambda-app/blob/main/site/package.json#L25C4-L25C34) by @gotgenes.

For anyone trying to get it working with React Router, I managed to do it by building the code with Vite normally using the React Router plugin, exporting an Express server as @curran  showed, and then building again using the bundling option from the NodejsFunction in CDK.

```
const ssr = new NodejsFunction(this, 'SsrServerFunction', {
    runtime: Runtime.NODEJS_20_X,
    entry: resolve(__dirname, '../../apps/web/serverless.js'),
    handler: 'handler',
    timeout: Duration.seconds(29),
    bundling: {
        format: OutputFormat.CJS,
        platform: 'node',
        target: 'node20',
        externalModules: ['@aws-sdk/*'],
    },
})
```


This was the only way I found, since Vite won’t export to CJS at all when using the plugin.

**Reply by MattyBalaam** [reply]
Why the need to stick with CJS? We have been using `OutputFormat.ESM` for a while now

**Reply by pimentellima** [reply]
> Why the need to stick with CJS? We have been using `OutputFormat.ESM` for a while now

couldnt get it to work with ESM on Lambda, maybe that's a skill issue

**Comment by ajaishankar** [comment]
Do look into SST

https://sst.dev/docs/start/aws/remix

For serverless, it creates a CloudFront distribution with static and dynamic behaviors.

1. Static files have a S3 origin
2. The application is packaged into a single lambda

**Reply by rreeves8** [reply]
I don't understand why SST exists, Can't you just use the aws SDK?

**Reply by ajaishankar** [reply]
For me, I could get a Remix app deployed in like 10 lines of SST config 😄 (as it already had a full stack setup)

They have a blog post that goes into detail on Pulumi, CDK, Cloud Formation etc.

https://sst.dev/blog/moving-away-from-cdk/

**Reply by corybill** [reply]
@rreeves8 I was a terraform guy. Never wanted to bother with all of the serverless frameworks. I recently tried SST and it is unbelievable. I'll never go back.

It streamlines local dev environment. Makes it consistent for all devs working on the project. With a push of a button I have my remix app deployed in lambdas to the edge with all the bells and whistles installed. It is f'in beautfiul.

**Reply by ajaishankar** [reply]
Do see the section "How does SST compare to CDK for Terraform or Pulumi?" here: https://sst.dev/docs/

> In a nutshell, SST is for developers, while CDKTF and Pulumi are primarily for DevOps engineers. 

[Live Lambda](https://sst.dev/docs/live#how-it-works) is really amazing.

Now SST supports [containers](https://sst.dev/docs/start/aws/remix#containers) too!

**Reply by rreeves8** [reply]
> @rreeves8 I was a terraform guy. Never wanted to bother with all of the serverless frameworks. I recently tried SST and it is unbelievable. I'll never go back.
> 
> It streamlines local dev environment. Makes it consistent for all devs working on the project. With a push of a button I have my remix app deployed in lambdas to the edge with all the bells and whistles installed. It is f'in beautfiul.

Sure man. 

Except sst doesn’t align with some companies compliance. We require terraform and have restrictions on what we can use. 

Not to say I couldn’t get this to work. But the framework does need better support, so people with limited brain cells like @JhonnyTheRipper can figure it out.

**Reply by rreeves8** [reply]
SST also doesn’t align with existing infrastructure. If your lambda is deployed using private networking or other infrastructure it’s another layer of complexity to manage.
